relayer/src/services/polygon-zkevm.ts

```diff
@@ -0,0 +1,322 @@
+  },
+};
+export class PolygonZKEvmRelayerService extends BaseRelayerService {
+  MERKLE_PROOF_ENDPOINT = "/merkle-proof";
+  BRIDGES_ENDPOINT = "/bridge";
+  customNetworkContracts: NetworkContracts;
+
+  async getNetworkConfig(): Promise<NetworkContracts> {
+    const l2NetworkId = await this.getL2ChainId();
+    return this.customNetworkContracts ?? CONTRACTS[l2NetworkId];
+  }
+
+  async getMessageMerkleProof(
+    depositCount: number,
+    originNetwork: number,
+  ): Promise<MerkleProofResponse | undefined> {
+    const networkConfig = await this.getNetworkConfig();
+
+    const merkleProofUrl = new URL(networkConfig.bridgeServiceUrl + this.MERKLE_PROOF_ENDPOINT);
+    merkleProofUrl.searchParams.set("deposit_cnt", depositCount.toString());
+    merkleProofUrl.searchParams.set("net_id", originNetwork.toString());
+
+    const response = await fetch(merkleProofUrl);
+    const responseJson = await response.json();
+
+    if (responseJson.code && responseJson.message) {
+      throw new Error(responseJson.message);
+    }
+
+    return responseJson;
+  }
+
+  async getMessageInfoSafe(
+    depositCount: number,
+    originNetwork: number /** 1 means Polygon ZkEVM (L2) and 0 means L1 */,
+  ): Promise<PolygonZKEvmDeposit> {
+    // Loop and retry until the endpoint returns a proper result;
+    try {
+      const response = await this.getMessageInfo(depositCount, originNetwork);
+      return response;
+    } catch (e) {
+      if (e.message == "not found in the Storage") {
+        await sleep(2000);
+        return this.getMessageInfoSafe(depositCount, originNetwork);
+      }
+      throw e;
+    }
+  }
+
+  async getMessageInfo(
+    depositCount: number,
+    originNetwork: number /** 1 means Polygon ZkEVM (L2) and 0 means L1 */,
+  ): Promise<PolygonZKEvmDeposit> {
+    const networkConfig = await this.getNetworkConfig();
+
+    const messageBridgeUrl = new URL(networkConfig.bridgeServiceUrl + this.BRIDGES_ENDPOINT);
+    messageBridgeUrl.searchParams.set("deposit_cnt", depositCount.toString());
+    messageBridgeUrl.searchParams.set("net_id", originNetwork.toString());
+
+    const response = await fetch(messageBridgeUrl);
+    const responseJson = await response.json();
+
+    if (responseJson.code && responseJson.message) {
+      throw new Error(responseJson.message);
+    }
+
+    return responseJson.deposit;
+  }
+
+  async prepare(): Promise<boolean> {
+    return true;
+  }
+
+  async relayTxToL1(l2TransactionHash: TransactionHash): Promise<string | undefined> {
+    console.log("\n");
+    console.log("Claiming PolygonZKEVM message on Ethereum L1.");
+
+    const receipt = await this.l2RpcProvider.waitForTransaction(l2TransactionHash, 1, 300000);
+
+    if (!receipt) {
+      throw new Error(`Transaction "${l2TransactionHash}" cannot be found on Polygon ZKEvm...`);
+    }
+    if (!receipt.status) {
+      throw new Error(`Transaction "${l2TransactionHash}" reverted on Polygon ZKEvm...`);
+    }
```

Comment:
This block is nearly the same as the block in finalize. How about we extract this into a separate function?

```
async function retrieveTransactionReceipt(provider: any, transactionHash: string) {
  const receipt = await provider.waitForTransaction(transactionHash, 1, 300000);

  if (!receipt) {
    throw new Error(`Transaction "${transactionHash}" cannot be found on L1.`);
  }
  if (!receipt.status) {
    throw new Error(`Transaction "${transactionHash}" reverted on L1.`);
  }

  return receipt;
}

```